src/Sequence.cc

```diff
@@ -304,15 +304,30 @@ void Sequence::execute(Context& context, CommChannel* comm)
     send_message(comm, Message::Type::sequence_started, "Sequence started",
                 Clock::now(), 0);
 
+    bool exception_thrown = false;
+    std::string exception_message;
+    StepIndex exception_index = 0;
```

Comment:
We loose information with this construct.

Imagine we could set it to `NaN` here instead.
Later at line 351 we send a `Message{}` with the index set. But is that information (the transported index) correct?
No one will ever know. If a 'normal' exception happened we send this initialized value. I.e. a 'normal' exception in Step number 3 still is reported as 'happened in index 0'.

So if there would be a `NaN` value possible the Message listener could be sure the `index` in the Message is correct.

First I thought we could just expand `exception_thrown` to be tri-state (no, yes-index-ok, yes-index-unknown), but that would probably not even help, because there is no `Message{}` where we can omit the `index`, or is there? No.

We could solve THAT by adding a `Message` ctor without `index`, that stores that there is no index available, and later on `get_index()` returns ... the unknown token, or refuses to return, or hmm...

A simple way to get an additional state like `NaN` into `StepIndex` is to dedicate the value 'all bits set' to that. In contrast to having a signed `index` and negative means not-valid, we would only loose one number instead of half of all numbers.

Further looking into it, there is no safeguard against wrap-around of the index; as far as my superficial look told me: The code will run, even if we have more Steps in the Sequence then there are numbers in StepIndex, but of course it will wrap.
Shall we prevent that?
If we use `0xFFFF == InvalidStepIndex` we have to prevent reaching even that number.


So my proposal is:

* Add a way to express `StepIndex` is invalid as `StepIndex invalid = -1`.
* Prevent index wrap around (i.e. hard limit number of steps, already in `Sequence::push_back()`)
* In above line initialize to `-1` (or better to the predefined `InvalidStepIndex` value (which is -1 ;)
* Correct stream output that it outputs `InvalidStepIndex` not as `0xFFFF`
* All the rest of the code will work as expected (well as far as I checked)

Reply:
Good point. I have now introduced `invalid_step_index` == 65535. There is no output of this index in the library, so the user code must check for it.

While at it, I have also extended `Sequence` accordingly (throw in `insert()` and `push_back()` if the sequence is already full, add `max_size()` member). There is also some related clean-up.

Reply:
Short comment from me: I would improve the code and add a new `class` (or `struct`?), for example also called `StepIndex` that encapsulates the behaviour for special cases (invalid index etc.) instead of using _magic numbers_ that are published allover the code. For me the previous "code" smells 😉 . Hence I would favour a cleaner solution with special operator definition that did the arithmetics (`operator+`, `operator-`, `operator++`, `operator--`. ...), member functions like `set_invalid()` and `is_invalid()`, and hide the implementation:

```c++
#ifndef STEP_INDEX_H_
#define STEP_INDEX_H_

#include "taskomat/Error.h"

namespace task {

class StepIndex {
    int idx_ = 0;
    bool is_invalid_ = false;
public:
    StepIndex(int idx = 0) {
        idx_ = idx;
    }
    ...

    StepIndex& operator+(const StepIndex& idx) {
        if (is_invalid_) throw Error("Index is invalid");
        idx_ += idx.idx_;
        return *this;
    }

    StepIndex& operator+(int idx) {
        if (is_invalid_) throw Error("Index is invalid");
        idx_ += idx;
        return *this;
    }
    ...

    void set_invalid(bool is_invalid) {
        is_invalid_ = is_invalid;
    }

    bool is_invalid() const {
        return is_invalid_;
    }
};

} // namespace task

#endif
```

It is more typing but future-safe.

Reply:
Well, there is no "magic number" in the usual sense in the code.

I have strong types for things like indices in some of the HLC libraries, but they come at a real cost. They *are* worth it if there is a real danger in confusing them with other parameters in function calls, but I do not think that is the case here. Sometimes simpler is better.

Reply:
> add a new `class` (or `struct`?), for example also called `StepIndex` that encapsulates the behaviour for special cases

Let's discuss that on Wednesday, OK? I am not really opposed, but I strongly think this should not become part of this PR. We can add it later if desired.

Reply:
Apart from the 'technical' aspect, there was no statement about signalling the wrong (i.e. `0`) `index` on 'irregular' exceptions. Reopening.

Reply:
I mean, using a `NaN` `StepIndex` is just one way to solve that problem. We could also add another `Message` member that communicates if we have an `index` or not. Also very clean solution, maybe more work.

Reply:
`Message` could even throw if the user did not check if there is an `index` or not and just tries to access it.
The more I think about that, the more I would even favor it above `NaN`.
(Like accessing an unfilled `optional`.)

Reply:
We *could* actually use an `optional<StepIndex>` in the output functions. I would be in favor of extending their signature for other reasons as well.

Reply:
As just discussed, I'll work on a solution using `optional<StepIndex>`. In the meantime I'll convert this into a draft.